content/4_Gold/BFS.mdx

```diff
@@ -254,6 +254,155 @@ deque.removeLast(); // [1, 3]
 
 ### Solution - Message Route
 
+**Time Complexity:** $\mathcal{O}(V+E)$
+
+We can observe is that there are many possible shortest paths to output. Fortunately, the problem states that we can print any valid solution. Notice that like every other 0/1 BFS problem, the distance of each node increases by 1 when we travel to the next level of unvisited nodes. However, the problem requires that we add additional information - in this case, the path. When we traverse from node $a$ to $b$, we can set the parent of $b$ to $a$. After the BFS is complete, this allows us to backtrack through the parents which ultimately leads us to our starting node. We know to terminate at node $1$ because its parent is $0$. If there is no path to our end node, then its distance will remain at [INT_MAX](https://en.cppreference.com/w/c/types/limits).
+
+For the test input, we start with the following parent array.
+
+<center>
+<table className='text-center'>
+<tbody>
+<tr>
+<td>Node</td>
+<td>1</td>
+<td>2</td>
+<td>3</td>
+<td>4</td>
+<td>5</td>
+</tr>
+<tr>
+<td>Parent</td>
+<td>0</td>
+<td>0</td>
+<td>0</td>
+<td>0</td>
+<td>0</td>
+</tr>
+<tr>
+<td>Distance</td>
+<td>0</td>
+<td>INT_MAX</td>
+<td>INT_MAX</td>
+<td>INT_MAX</td>
+<td>INT_MAX</td>
+</tr>
+</tbody>
+</table>
+</center>
+
+After visiting children of node $1$:
+
+<center>
+<table className='text-center'>
+<tbody>
+<tr>
+<td>Node</td>
+<td>1</td>
+<td>2</td>
+<td>3</td>
+<td>4</td>
+<td>5</td>
+</tr>
+<tr>
+<td>Parent</td>
+<td>0</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td>0</td>
+</tr>
+<tr>
+<td>Distance</td>
+<td>0</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td className="bg-green-200 dark:bg-green-800">1</td>
+<td>INT_MAX</td>
+</tr>
+</tbody>
+</table>
+</center>
+
+After visiting node $5$ from node $4$
+
+<center>
+<table className='text-center'>
+<tbody>
+<tr>
+<td>Node</td>
+<td>1</td>
+<td>2</td>
+<td>3</td>
+<td>4</td>
+<td>5</td>
+</tr>
+<tr>
+<td>Parent</td>
+<td>0</td>
+<td>1</td>
+<td>1</td>
+<td>1</td>
+<td className="bg-green-200 dark:bg-green-800">4</td>
+</tr>
+<tr>
+<td>Distance</td>
+<td>0</td>
+<td>1</td>
+<td>1</td>
+<td>1</td>
+<td className="bg-green-200 dark:bg-green-800">2</td>
+</tr>
+</tbody>
+</table>
+</center>
+
+To determine the path, we can backtrack from node $n \rightarrow 1$, in this case $5 \rightarrow 1$, pushing each value that we back track into a vector. The path we take is $5 \rightarrow \text{parent}[5]=4 \rightarrow \text{parent}[4] =1$ which corresponds to the vector $[5, 4, 1]$. We break at node 1 because we notice that its parent is 0. Finally, we reverse the vector and print out it's length (in this case, 3).
```

Comment:
```suggestion
To determine the path, we can backtrack from node $n \rightarrow 1$, in this case $5 \rightarrow 1$, pushing each value that we back track into a vector. The path we take is $5 \rightarrow \text{parent}[5]=4 \rightarrow \text{parent}[4] =1$ which corresponds to the vector $[5, 4, 1]$. We break at node $1$ because it was the initial starting node. Finally, we reverse the vector and print out it's length (in this case, $3$).
```